mat/band.go

```diff
@@ -136,6 +136,12 @@ func NewBandDense(r, c, kl, ku int, data []float64) *BandDense {
 	}
 }
 
+// NewDiagonal is a convenience function that returns a diagonal matrix represented by a
+// BandDense. The length of data must be min(r, c) otherwise NewDiagonal will panic.
+func NewDiagonal(r, c int, data []float64) *BandDense {
```

Comment:
Yes, I agree with the intent. Wikipedia suggests that the term Diagonal "usually refers to square matrices", and calls non-square ones "rectangular diagonal matrices". 

I would go with `NewDiagonal` for the symmetric case, and `NewDiagonalBand` or `NewDiagonalRect` for the rectangular case.

`data` should be allowed to be nil, just like the other `NewXxx` functions. I commonly create a new matrix then use a loop to set the diagonal elements (and fortunately these will be much more efficient soon).